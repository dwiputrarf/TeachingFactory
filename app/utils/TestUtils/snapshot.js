/* eslint-disable import/no-extraneous-dependencies */
import renderer from 'react-test-renderer';

const mockSetState = Component =>
  class extends Component {
    setState(f, callback) {
      if (typeof f === 'function') {
        /* eslint-disable react/no-direct-mutation-state */
        this.state = { ...this.state, ...f(this.state) };
      } else {
        /* eslint-disable react/no-direct-mutation-state */
        this.state = { ...this.state, ...f };
      }
      if (callback) callback();
    }
  };

const assertSnapshot = (C, config) => {
  const { props, state, desc, throws, beforeTest } = config;

  let Component;
  let RenderedComponent;

  if (beforeTest) {
    beforeTest();
  }

  if (throws) {
    test(desc, () => {
      expect(() => {
        if (C.prototype.render) {
          Component = new C(props);
        } else {
          Component = C(props);
        }
      }).toThrow(throws);
    });
  } else {
    if (C.prototype.render) {
      const CWithMockState = mockSetState(C);
      Component = new CWithMockState(props);

      if (state) {
        Component.setState(state);
      }

      RenderedComponent = Component.render();
    } else {
      RenderedComponent = C(props);
    }

    test(desc, () => {
      const tree = RenderedComponent && renderer.create(RenderedComponent).toJSON();
      expect(tree).toMatchSnapshot();
    });
  }
};

const assertSnapshots = (Component, configs) => {
  configs.forEach(config => assertSnapshot(Component, config));
};

export { assertSnapshots, mockSetState };
export default { assertSnapshots, mockSetState };
