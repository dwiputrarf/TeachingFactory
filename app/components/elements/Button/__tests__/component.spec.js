import { assertSnapshots, mockSetState } from '../../../../utils/TestUtils/snapshot';
import Component from '../component';

describe('Component snapshot testing', () => {
  const configs = [
    {
      state: {},
      props: { type: 'raised' },
      desc: 'renders with type: raised'
    },
    {
      state: {},
      props: { type: 'flat' },
      desc: 'renders with type: flat'
    },
    {
      state: {},
      props: { type: 'flat-ripple' },
      desc: 'renders with type: flat-ripple'
    },
    {
      state: {},
      props: { type: 'raised', disabled: true },
      desc: 'renders with disabled'
    }
  ];
  assertSnapshots(Component, configs);
});

describe('Component', () => {
  const ComponentWithMockedState = mockSetState(Component);
  const getInstance = props => new ComponentWithMockedState(props);
  beforeEach(() => {
    this.props = {
      type: 'raised',
      navigation: {
        navigate: jest.fn()
      }
    };
  });

  afterEach(() => {
    this.props = null;
    jest.clearAllMocks();
  });

  test('render on android', () => {
    jest.mock('Platform', () => {
      const Platform = require.requireActual('Platform');
      Platform.OS = 'android';
      return Platform;
    });
    const instance = getInstance(this.props);
    instance.render();
  });

  test('render on ios', () => {
    jest.mock('Platform', () => {
      const Platform = require.requireActual('Platform');
      Platform.OS = 'ios';
      return Platform;
    });
    const instance = getInstance(this.props);
    instance.render();
  });
});
