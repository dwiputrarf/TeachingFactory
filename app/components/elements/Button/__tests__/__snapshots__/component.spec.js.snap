// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component snapshot testing renders with disabled 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#6DD9D2",
      "borderRadius": 5,
      "height": 40,
      "justifyContent": "center",
      "opacity": 1,
      "width": 200,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#6DD9D2",
          "borderRadius": 5,
          "height": 40,
          "justifyContent": "center",
          "width": 200,
        },
        undefined,
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#bdbdbd",
          },
          undefined,
        ]
      }
    />
  </View>
</View>
`;

exports[`Component snapshot testing renders with type: flat 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "borderRadius": 5,
      "height": 40,
      "justifyContent": "center",
      "opacity": 1,
      "width": 200,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "borderRadius": 5,
          "height": 40,
          "justifyContent": "center",
          "width": 200,
        },
        undefined,
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#bdbdbd",
          },
          undefined,
        ]
      }
    />
  </View>
</View>
`;

exports[`Component snapshot testing renders with type: flat-ripple 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "borderRadius": 5,
      "height": 40,
      "justifyContent": "center",
      "opacity": 1,
      "width": 200,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "borderRadius": 5,
          "height": 40,
          "justifyContent": "center",
          "width": 200,
        },
        undefined,
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#bdbdbd",
          },
          undefined,
        ]
      }
    />
  </View>
</View>
`;

exports[`Component snapshot testing renders with type: raised 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#289F97",
      "borderRadius": 5,
      "height": 40,
      "justifyContent": "center",
      "opacity": 1,
      "width": 200,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#289F97",
          "borderRadius": 5,
          "height": 40,
          "justifyContent": "center",
          "width": 200,
        },
        undefined,
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#bdbdbd",
          },
          undefined,
        ]
      }
    />
  </View>
</View>
`;
